src/parser.rs

```diff
@@ -166,51 +155,244 @@ impl<'a> DocParser<'a> {
+      self.collect_diagnostics_for_nodes(doc_nodes);
+    }
+
+    Ok(doc_nodes_by_url)
+  }
+
+  fn resolve_references_for_nodes(
+    &self,
     specifier: &ModuleSpecifier,
-  ) -> Result<Vec<DocNode>, DocError> {
-    let module_info = self.get_module_info(specifier)?;
-    let doc_nodes = self.get_doc_nodes_for_module_info(module_info)?;
-    self.collect_diagnostics_for_nodes(&doc_nodes);
-    Ok(doc_nodes)
+    nodes: &mut Vec<DocNode>,
+    name_path: &[String],
+    all_locations: &mut HashSet<Location>,
+  ) -> Result<(), DocError> {
+    let mut i = 0;
+    while i < nodes.len() {
+      let name = nodes[i].name.to_string();
+
+      match &mut nodes[i].def {
+        DocNodeDef::Namespace { namespace_def } => {
+          let mut namespace_elements = namespace_def
+            .elements
+            .iter()
+            .map(|node| Rc::unwrap_or_clone(node.clone()))
+            .collect::<Vec<_>>();
+
+          let mut name_path = name_path.to_vec();
+          name_path.push(name);
+
+          self.resolve_references_for_nodes(
+            specifier,
+            &mut namespace_elements,
+            &name_path,
+            all_locations,
+          )?;
+
+          namespace_def.elements =
+            namespace_elements.into_iter().map(Rc::new).collect();
+        }
+        DocNodeDef::Reference { reference_def } => {
+          let mut name_path = name_path.to_vec();
+          name_path.push(name);
+
+          if !all_locations.contains(&reference_def.target) {
+            if let Some(new_nodes) = self.resolve_dangling_reference(
+              specifier,
+              reference_def,
+              name_path,
+              false,
+            )? {
+              nodes.splice(i..=i, new_nodes);
+              all_locations
+                .extend(nodes.iter().map(|node| node.location.clone()));
+              continue;
+            }
+          }
+        }
+
+        _ => {}
+      }
+
+      i += 1;
+    }
+    Ok(())
   }
 
-  /// Fetches `file_name`, parses it, and resolves its reexports.
-  pub fn parse_with_reexports(
+  fn resolve_dangling_reference(
     &self,
     specifier: &ModuleSpecifier,
-  ) -> Result<Vec<DocNode>, DocError> {
-    let doc_nodes =
-      self.parse_with_reexports_inner(specifier, HashSet::new())?;
-    self.collect_diagnostics_for_nodes(&doc_nodes);
-    Ok(doc_nodes)
+    reference_def: &ReferenceDef,
+    mut name_path: Vec<String>,
+    star: bool,
+  ) -> Result<Option<Vec<DocNode>>, DocError> {
+    let module = resolve_deno_graph_module(self.graph, specifier)?;
+
+    match module {
+      Module::Js(_) | Module::Json(_) | Module::Wasm(_) => {
+        let module_info = self.get_module_info(module.specifier())?;
+        let exports = module_info.exports(&self.root_symbol);
+        let root_name = name_path.remove(0);
+        let (mut export_name, export) = exports
+          .resolved
+          .iter()
+          .find(|(name, _)| &&root_name == name)
+          .unwrap();
```

Comment:
Maybe let's not unwrap for this one? It could cause issues for bad references?

Reply:
how could there be a bad reference though?

Reply:
I guess it's not possible.